vultr/resource_block_storage.go

```diff
@@ -0,0 +1,169 @@
+				Computed: true,
+			},
+
+			"label": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"dcid": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"size_gb": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+
+			"date_created": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"cost_per_month": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"attached_to_subid": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceBlockStorageCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Client)
+
+	label := d.Get("label").(string)
+	dcid := d.Get("dcid").(int)
+	sizeGb := d.Get("size_gb").(int)
+
+	log.Printf("[INFO] Creating new block storage")
+	storage, err := client.CreateBlockStorage(label, dcid, sizeGb)
+	if err != nil {
+		return fmt.Errorf("Error creating block storage: %v", err)
+	}
+	d.SetId(storage.ID)
+
+	old, new := d.GetChange("attached_to_subid")
+	if old != new {
+		if err := client.AttachBlockStorage(d.Id(), new.(string)); err != nil {
+			return fmt.Errorf("Error attaching block storage (%s) to %q: %v", d.Id(), new.(string), err)
+		}
+	}
+
+	return resourceBlockStorageRead(d, meta)
+}
+
+func resourceBlockStorageRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Client)
+
+	storage, err := client.GetBlockStorage(d.Id())
+	if err != nil {
+		if err.Error() == "Invalid block storage." {
+			log.Printf("[WARN] Removing block storage (%s) because it is gone", d.Id())
+			d.SetId("")
+			return nil
+		}
+		d.SetId("")
+		return fmt.Errorf("Error getting block storage (%s): %v", d.Id(), err)
+	}
+
+	d.Set("cost_per_month", storage.Cost)
+	d.Set("size_gb", storage.SizeGB)
+	d.Set("label", storage.Name)
+	d.Set("dcid", storage.RegionID)
+	d.Set("status", storage.Status)
+	d.Set("attached_to_subid", storage.AttachedTo)
+	d.Set("subid", storage.ID)
+
+	return nil
+}
+
+func resourceBlockStorageUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Client)
+
+	d.Partial(true)
+
+	if d.HasChange("label") {
+		log.Printf("[INFO] Updating block storage (%s) name", d.Id())
+		_, new := d.GetChange("label")
+		if err := client.LabelBlockStorage(d.Id(), new.(string)); err != nil {
+			return fmt.Errorf("Error renaming block storage (%s) to %q: %v", d.Id(), new.(string), err)
+		}
+		d.SetPartial("label")
+	}
+
+	if d.HasChange("size_gb") {
+		log.Printf("[INFO] Resizing block storage (%s)", d.Id())
+		_, new := d.GetChange("size_gb")
+		if err := client.ResizeBlockStorage(d.Id(), new.(int)); err != nil {
+			return fmt.Errorf("Error resizing block storage (%s) to %q: %v", d.Id(), new.(int), err)
+		}
+		d.SetPartial("size_gb")
+	}
+
+	if d.HasChange("attached_to_subid") {
+		log.Printf("[INFO] Attaching/detaching block storage (%s)", d.Id())
```

Comment:
Rather than logging just once, I think it would be clearer to log for each actual step.